src/brod_producer.erl

```diff
@@ -16,493 +16,282 @@
-    {ok, State} ->
-      {noreply, State};
-    {error, Reason} ->
-      %% Errors found in produce response, terminate myself.
-      %% In a nicer way, it should recognize at least error code
-      %% ?EC_NOT_LEADER_FOR_PARTITION, then try to connect to the
-      %% new leader (if not yet connected) and migrate data buffer
-      %% and pending list from the old leader to the new leader.
-      %% We can not do this now because the payload is not kept in
-      %% the pending list for resend, there is no better solution
-      %% than crash myself and hope the brod clients would (and they
-      %% should) retry after restart.
-      {stop, Reason, State0}
+handle_info(init_socket, #state{ client_id = ClientId
+                               , topic     = Topic
+                               , partition = Partition
+                               , config    = Config0
+                               } = State) ->
+  %% 1. Fetch and validate metadata
+  {ok, Metadata} = brod_client:get_metadata(ClientId, Topic),
+  #metadata_response{ brokers = Brokers
+                    , topics  = [TopicMetadata]
+                    } = Metadata,
+  #topic_metadata{ error_code = TopicErrorCode
+                 , partitions = Partitions
+                 } = TopicMetadata,
+  brod_kafka:is_error(TopicErrorCode) andalso
+    erlang:throw({"topic metadata error", TopicErrorCode}),
+  #partition_metadata{ error_code = PartitionErrorCode
+                     , leader_id  = LeaderId} =
+    lists:keyfind(Partition, #partition_metadata.id, Partitions),
+  brod_kafka:is_error(PartitionErrorCode) andalso
+    erlang:throw({"partition metadata error", PartitionErrorCode}),
+  LeaderId >= 0 orelse
+    erlang:throw({"no leader for partition", {ClientId, Topic, Partition}}),
+  #broker_metadata{host = Host, port = Port} =
+    lists:keyfind(LeaderId, #broker_metadata.node_id, Brokers),
+
+  %% 2. Lookup, or maybe (re-)establish a connection to partition leader
+  {ok, SockPid} = brod_client:get_connection(ClientId, Host, Port),
+  _ = erlang:monitor(process, SockPid),
+
+  %% 3. Register self() to client.
+  ok = brod_client:register_producer(ClientId, Topic, Partition),
+
+  %% 4. Create the producing buffer
+  {BufferLimit, Config1} = take_config(partition_buffer_limit,
+                                       ?DEFAULT_PARITION_BUFFER_LIMIT,
+                                       Config0),
+  {OnWireLimit, Config2} = take_config(partition_onwire_limit,
+                                       ?DEFAULT_PARITION_ONWIRE_LIMIT,
+                                       Config1),
+  {MsgSetBytes, Config}  = take_config(message_set_bytes_limit,
+                                       ?DEFAULT_MESSAGE_SET_BYTES_LIMIT,
+                                       Config2),
+  RequiredAcks = get_required_acks(Config),
+  AckTimeout = get_ack_timeout(Config),
+  SendFun =
+    fun(KafkaKvList) ->
+      %% TODO: change to plain kv-list ?
```

Comment:
yes, if we stick to per-partition producer, and all the requests are always set per-partition, dict is kind of a cumbersome data structure since a plain list is enough.